src/core/thread/mle_router.cpp

```diff
@@ -1284,6 +1284,7 @@ otError MleRouter::HandleAdvertisement(const Message &         aMessage,
         if (processRouteTlv)
         {
             SuccessOrExit(error = ProcessRouteTlv(route));
+            aNeighbor = NULL; // aNeighbor is no longer valid after `ProcessRouteTlv`
```

Comment:
Can you please explain the reason for this?

 We have check with `aNeighbor` below for example at `exit` which would not happen (if we go through this code path). Do we know if this would be ok?
```
if (aNeighbor && aNeighbor->GetRloc16() != sourceAddress)
    {
        // Remove stale neighbors
        RemoveNeighbor(*aNeighbor);
    }
```

Reply:
The original code has bug in it:

Let's assume that `aNeighbor` references to a `Router` instance in `mRouterTable`. 
- After `ProcessRouteTlv` is executed, `aNeighbor`  can point to a different `Router`, because `ProcessRouteTlv` can move Routers in `mRouterTable`. So, after `ProcessRouteTlv`, the `Router` instance referenced by `aNeighbor` is not what we expect it to be, the sender of the `MLE Advertisement`. So, any later handling of `aNeighbor` can be a potential bug. 
- If we are not lucky, `aNeighbor` now references to a `Router` instance with `RLOC16 == 0xFFFF` (Router ID not allocated). In such case, `RemoveNeighbor(*aNeighbor);` at the `exit` will be executed and the original `RemoveNeighbor` will treat this neighbor as a `Child` instance which it is not. 

One way of addressing it would be reassigning `aNeighbor` to the correct `Router`. Meanwhile assigning `aNeighbor` to `NULL` seems also OK to me according to analysis of the following code in this function.

Reply:
Thanks.

Reply:
This PR is to fix issues found in a 8x8 nodes simulation. The bugs are subtle and difficult to comprehend and unclear on how to handle. The current code changes are more for discussion than for fixing.

Reply:
This is indeed a good find! 👍

Reply:
Setting `aNeighbor = NULL` is probably sufficient here. Wherever we need a router entry below, we perform another lookup.